src/aggregation.jl

```diff
@@ -341,8 +345,17 @@ internally.
 
 *Performs no argument validation, caller should do this.*
 """
-_inverse_eltype_tuple(ts::NTransforms, ys::Tuple) =
-    reduce(promote_type, map(inverse_eltype, ts, ys))
+@generated function _inverse_eltype_tuple(ts::NTransforms{N}, ::Type{T}) where {N,T}
+    ex = mapreduce((i, T) -> :(inverse_eltype(ts[$i], $T)),
+                   (a, b) -> :(promote_type($a, $b)),
+                   1:N, fieldtypes(T))
+    quote
+        A = ts
+        B = fieldtypes(T)
+        @argcheck N == length(B) "Incompatible input length."
+        $(ex)
+    end
+end
```

Comment:
I'm not a big fan of generated functions (but sometimes it's unavoidable) - I wonder if a recursive definition is inferred as well, something like
```suggestion
function _inverse_eltype_tuple(ts::NTransforms{N}, ::Type{T}) where {N,T<:Tuple}
    @argcheck T <: NTuple{N,Any} "Incompatible input length."
    return __inverse_eltype_tuple(ts, T)
end
function __inverse_eltype_tuple(ts::NTransform, ::Type{Tuple{}})
   return Union{}
end
function __inverse_eltype_tuple(ts::NTransform, ::Type{T}) where {T<:Tuple}
   return promote_type(inverse_eltype(Base.first(ts), fieldtype(T, 1)), __inverse_eltype_tuple(Base.tail(ts), Tuple{Base.tail(fieldtypes(T))...}))
end
```